src/arch/cuda/hybridAPI/buddy_allocator.h

```diff
@@ -0,0 +1,247 @@
+    // to track (internal) fragmentation. Does not contain the pointer as it
+    // is stored as a key in the unordered map.
+    struct AllocBlock {
+      size_t size;
+      size_t requested;
+
+      AllocBlock() : size(0), requested(0) {}
+
+      AllocBlock(size_t size_, size_t requested_) : size(size_), requested(requested_) {}
+    };
+
+    // Allocation limits
+    size_t limit;
+    const size_t min_alloc;
+
+    // Base pointer of the initial allocation
+    uint8_t* base_ptr;
+
+    // Buckets each with a free list
+    std::list<FreeBlock>* buckets;
+    int bucket_count;
+
+    // Map of allocated blocks
+    std::unordered_map<uint8_t*, AllocBlock> alloc_map;
+
+    /* --------------------- */
+    /* | Utility functions | */
+    /* --------------------- */
+
+    void print_status() {
+      printf("(buckets)\n");
+      size_t free = 0;
+      for (int i = 0; i < bucket_count; i++) {
+        printf("bucket[%d]: ", i);
+        for (const auto& block : buckets[i]) {
+          free += block.size;
+          printf("{%p, %lu} ", block.ptr, block.size);
+        }
+        printf("\n");
+      }
+
+      printf("(alloc_map)\n");
+      size_t allocated = 0;
+      size_t used = 0;
+      for (const auto& elem : alloc_map) {
+        const auto& block = elem.second;
+        allocated += block.size;
+        used += block.requested;
+        printf("ptr: %p, size: %lu, req: %lu\n", elem.first, block.size, block.requested);
+      }
+
+      printf("(fragmentation) free: %lu, allocated: %lu, used: %lu\n", free, allocated, used);
+    }
+
+    size_t get_free_size() {
+      size_t free = 0;
+      for (int i = 0; i < bucket_count; i++) {
+        for (const auto& block : buckets[i]) {
+          free += block.size;
+        }
+      }
+
+      return free;
+    }
+
+    int get_bucket(size_t size) {
+      return (int)std::ceil(std::log2((double)size)) - 2;
+    }
+
+    int get_block_index(uint8_t* ptr, size_t size) {
+      return (size_t)(ptr - base_ptr) / size;
+    }
+
+    /* ------------------------ */
+    /* | Allocation functions | */
+    /* ------------------------ */
+
+    allocator(size_t size) : min_alloc(4), base_ptr(NULL) {
+      if (size <= 0) {
+        fprintf(stderr, "Allocator size has to be larger than 0 bytes\n");
+        abort();
+      }
+
+      // Request GPU memory (closest power of 2)
+      int limit_log2 = std::ceil(std::log2((double)size));
+      limit = (size_t)std::pow(2, limit_log2);
+      cudaMalloc(&base_ptr, limit);
+      DEBUG_PRINT("Initialized base_ptr %p with %lu bytes\n", (void*)base_ptr, limit);
+
+      // Initialize buckets and set up last bucket (for size min_alloc)
+      bucket_count = limit_log2 - 1;
+      buckets = new std::list<FreeBlock>[bucket_count];
+      buckets[bucket_count-1].emplace_back(base_ptr, limit);
+    }
+
+    ~allocator() {
+      // Free GPU memory
+      cudaFree(base_ptr);
+      delete[] buckets;
+    }
+
+    void* malloc(size_t request) {
```

Comment:
I would suggest changing the name of `request` to `requestedSize` or something that gives and an indication that it is a size of some sort. The same suggestion for `limit`. (And you can change the `size` parameter in `allocator`'s constructor to maybe `deviceAllocationSize`).

Reply:
Additionally, based on the suggestion for variable names in the style guide (https://github.com/UIUC-PPL/charm/wiki/Charm-Runtime-Style-Guide), it'll be good to skip the underscore in variable names and use the capitalization format.

Reply:
Agreed on `request` and `limit`. But as for the style, HAPI is mostly based on the Google style (at least my additions), so will keep it that way for now.